### ISSUE TITLE
Isthmus: Improve the API for Substrait to Sql

### ISSUE THREAD

**Issue by mbwhite** [substrait-io/substrait-java#276]
Background: this came out of our recent work using Substrait - and found that converting to SQL was bit cumbersome. I've outlined a suggestion below based on the code we've got. The implementation is the first pass and fully expect to get suggestions for improvement.  More than happy to submit a PR for this and see it through to completion. Thanks!

**SubstraitToSql**
 This currently works at the Relation level not the Plan Level. As such the final SQL statement doesn't contain the expected final names of the columns.  What was a `select id as identifier` is lost to become a `select id as $f1` 

The proposal here is work at the plan level where this information can be added via modifying the final relational. The information is in the plan and therefore could be used. 

This would give symmetry to the APIs - and make them easier to use externally. Having to drop one level in the proto structure is cumbersome.

The downside of the approach is that in some cases calcite is creating an additional project relation. Have attempted to rationalise this via optimisers, but it felt better to leave the optimisation out.  And get an SQL statement that was correct.

_On the API structure.._

Should the API for Isthmus be `io.substrait.plan.Plan` or `io.substrait.proto.Plan`.  As this is an external API, then it would be useful to have consistency - especially with the return types that can't be overloaded? SqltoSubstrait is currently returning the protobuf versions.  Is there a reason that Isthmus does not return the `io.substrait.plan.Plan` versions - this would seem to be the higher-level API?

### Code

Add to the `SubtraitToCalcite` class
```
   /** Same as the convert(rel) method, but also takes a list of the intendedFieldNames
    * the top relation is modified to use these names
    */
    public RelNode convert(Rel rel, List<String> intendedFieldNames) {
        CalciteSchema rootSchema = toSchema(rel);
        RelBuilder relBuilder = createRelBuilder(rootSchema);
        SubstraitRelNodeConverter converter = createSubstraitRelNodeConverter(relBuilder);

        var calciteRelNode = rel.accept(converter);
        List<String> relationFieldNames = calciteRelNode.getRowType().getFieldNames();

        if (intendedFieldNames.size() != relationFieldNames.size()) {
            // no idea what this means, but surely means something is wrong
            throw new RuntimeException("Plan names different size to relation names");
        }
        List<RexNode> newAlias = new ArrayList<RexNode>();
        relBuilder = relBuilder.push(calciteRelNode);
        for (var x = 0; x < intendedFieldNames.size(); x++) {
            var alias = relBuilder.alias(
                    relBuilder.field(relationFieldNames.get(x)), intendedFieldNames.get(x));
            newAlias.add(alias);
        }

        var updatedCalciteNode = relBuilder.project(newAlias, ImmutableList.of(), true).build();

        return updatedCalciteNode;
    }
```

In SubstraitToSql, this new `convert()` is then called by an updated `toSql`
```
    public static List<String> toSql(Plan plan, SqlDialect sqlDialect) {
        try {
            var sqlStrings = new ArrayList<String>();

            SimpleExtension.ExtensionCollection extensions = SimpleExtension.loadDefaults();
            var converter = new SubstraitToCalcite_II(
                    extensions, new JavaTypeFactoryImpl(SubstraitTypeSystem.TYPE_SYSTEM));

            for (var root : plan.getRoots()) {

                List<String> intendedFieldNames = root.getNames();
                var calciteRelNode = converter.convert(root.getInput(), intendedFieldNames);
                // swap to the following line to see the difference 
                // var calciteRelNode = converter.convert(root.getInput());

                var sqlString = SubstraitToSql.toSql(calciteRelNode, sqlDialect);

                sqlStrings.add(sqlString);
            }

            return sqlStrings;
        } catch (IOException e) {
            throw new RuntimeException("Unable to convert to SQL", e);
        }
    }
  ```
  
  Externally this would look like 
```
            var substrait_proto_a = new SqlToSubstrait().execute(sqlQuery, Arrays.asList(createTables));
            var plan = new ProtoPlanConverter().from(substrait_proto_a);           
            List<String> sqlStrings = SubstraitToSql.toSql(plan);
            sqlStrings.forEach(System.out::println);
```

Notes:
- as mentioned what is the intention with return types  `io.substrait.proto` or the high-level `io.substrait.plan` calls?
- Error handling can be matched to the preferred handling in the isthmus library
- Should the style of APIs be match - static vs instance?

**Comment by vbarua** [comment]
> Is there a reason that Isthmus does not return the io.substrait.plan.Plan versions - this would seem to be the higher-level API?
 
It predates my time working with the library, but it's been on my list of thing to address. I don't think there's a good reason for `SqlToSubstrait` to return protos, I suspect it was implemented early on and never revisited. In my opinions Isthmus should work with the Substrait POJOs (i.e. `io.substrait.plan.Plan`) and we should excise any references to the protos that we find in it.

As for the API in `SubstraitToCalcite`, I think the convert method should actually return a `org.apache.calcite.rel.RelRoot` and not just a `org.apache.calcite.rel.RelNode`. Something like:
```java
public RelRoot convert(Root plan) {
```
From the RelRoot [docs](https://calcite.apache.org/javadocAggregate/org/apache/calcite/rel/RelRoot.html), the `fields` field can be used to assign output names for aliasing purposes.

The SubstraitRelVisitor (which converts from Calcite to Substrait) also does the wrong thing here IMO and throws away the RelRoot information:
https://github.com/substrait-io/substrait-java/blob/8537dca93b410177f2ee5aefffe83f7c02a3668c/isthmus/src/main/java/io/substrait/isthmus/SubstraitRelVisitor.java#L382-L389

These should return a Substrait `Root` object.

> Should the style of APIs be match - static vs instance?

I think it should be instance based, because I would like to see more configuration for the conversion.

I'm actually not super happy with the SQL conversion API as is, and aside from it being useful as an internal utility for fast testing I'm not sure how useful it is generally. Core Isthmus can handle converting from Calcite into Substrait (via the poorly named SubstraitRelVisitor) or from Substrait to Calcite (using SubstraitToCalcite). The SQL stuff feels tacked on and it's also not specific to Isthmus (or Substrait) because it should (ideally) only depend on Calcite if we layer the APIs correctly.

**Comment by mbwhite** [comment]
Thanks @vbarua - admit that I would agree with your views and opinions; things aren't quite lined-up. It will have evolved overtime as the standards and requirements have emerged. 

I'll look in more detail at the `SubstraitRelVisitor`

**Comment by nielspardon** [comment]
I did add a new method `RelRoot convert(Plan.Root root)` to `SubstraitToCalcite` recently via https://github.com/substrait-io/substrait-java/pull/339 which should address some but not all of the issues mentioned here.

**Comment by nielspardon** [comment]
The API improvements requested here have all landed across several PRs since this was filed. Mapping each point to the current state on `main`:

1. **`SubstraitToSql` should operate at the plan level so final column aliases survive** (`select id as identifier` no longer becoming `$f1`) — ✅ Done. `SubstraitToSql.convert(Plan, SqlDialect)` iterates `plan.getRoots()` and applies the final projection with the `Plan.Root` names, preserving aliases.
2. **APIs should use the `io.substrait.plan.Plan` POJO rather than `io.substrait.proto.Plan`** — ✅ Done. `SqlToSubstrait.convert(...)` now returns `io.substrait.plan.Plan`, and there are no proto-`Plan` references left in these entry points.
3. **`SubstraitToCalcite.convert()` should return a `RelRoot`** — ✅ Done in #339 (`convert(Plan.Root)` returns `RelRoot`).
4. **`SubstraitRelVisitor.convert()` should return a Substrait `Root` instead of discarding the `RelRoot` info** — ✅ Done; `convert(RelRoot, ConverterProvider)` returns `Plan.Root` and extracts the output names (with nested-type handling). The old `RelRoot`-discarding overloads are deprecated.
5. **Instance-based, configurable APIs rather than static** — ✅ Done. `SqlToSubstrait`, `SubstraitToSql`, and `SubstraitToCalcite` are instance-based and configured via `ConverterProvider`.

Closing as resolved.

A couple of follow-ups remain that are out of scope for this issue and better tracked separately — please open new issues if they're wanted:

- **`WriteRel` / DML-DDL round-tripping** through `SubstraitToCalcite` (currently flagged with a `TODO`; blocked on core `WriteRel` support).
- **Hierarchical (nested) field-name aliasing**, which is limited by Apache Calcite rather than by the Isthmus API.
- The earlier suggestion that the SQL-parsing utility layer ideally depend only on Calcite (and live outside Isthmus) — never actioned, and worth its own discussion if still desired.

**Comment by nielspardon** [comment]
Follow-up note: the third item above — making the SQL layer depend only on Calcite / providing just a reference SQL parser + generator — is already tracked in #362. The remaining two follow-ups (`WriteRel` / DML-DDL round-tripping and nested field-name aliasing) are still un-tracked and would need their own issues if wanted.
